rust/kernel/chrdev.rs

```diff
@@ -20,10 +19,57 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+/// Character device.
+///
+/// # Invariants
+///
+/// The pointer [`Cdev::cdev`] is valid and non-null.
+struct Cdev(*mut bindings::cdev);
+
+impl Cdev {
+    fn alloc(
+        fops: &'static bindings::file_operations,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult<Self> {
+        // SAFETY: `bindings::cdev_alloc` returns a valid pointer or null.
```

Comment:
Can you tell me more about Contexts? Are you saying that every function in Rust needs a `# Context` comment which indicates whether or not it can run in atomic context? Does kernel Rust have atomic contexts yet? A grep for `GFP_ATOMIC` brings up nothing.

Reply:
Only those that cannot be called from all contexts. This is one of the policies we introduced very recently and we do not really follow it in many places, so bear with us...

Furthermore, we do not require documenting private items like this struct (at least so far, still thinking about that one...), but callers that are public still need to be documented. So e.g. `Registration::register()` should have it.

Reply:
The long-term idea is to have some analyzer or lint tell us if we are following the rules by annotating the functions somehow -- or even the compiler itself (e.g. something like the "colored `unsafe`" I proposed).

Reply:
Thanks for clarifying, that makes sense! From what I understand, only a relatively small subset of kernel functions can be called from atomic contexts. So it might make sense to add a documentation marker (or real attribute?) only to those functions that are atomic compatible.

I can't think of any reason why one would want to create/register a `cdev` from an atomic context. But it's an interesting area to think about, for sure!

Reply:
PS looks like our messages crossed :)

Reply:
Yeah, when making things like this so strict, it brings up situations that do not make much sense (like registering, as you say). Similar to the `SAFETY` comments, it is more about being consistent, even if it should never be the case that someone attempts to register something in an atomic context etc.

If we manage to get the `# Context` comments right, it is my hope that reasoning about contexts is easier compared to the C side, plus we would be in a much better position later on to switch to actual annotations and automated checking.

Reply:
As for the `# Context` comment, you can ignore it for the moment -- we still need to see how we will write them and do a pass over the current code (we can discuss it in one of the meetings).

Reply:
Sounds good! I'll push the updated `SAFETY` comments in a few minutes.